test/server/admin/stats_handler_test.cc

```diff
@@ -50,6 +52,114 @@ INSTANTIATE_TEST_SUITE_P(IpVersions, AdminStatsTest,
+            "c2: 20\n"
+            "h1: P0(200.0,200.0) P25(202.5,202.5) P50(205.0,205.0) P75(207.5,207.5) "
+            "P90(209.0,209.0) P95(209.5,209.5) P99(209.9,209.9) P99.5(209.95,209.95) "
+            "P99.9(209.99,209.99) P100(210.0,210.0)\n"
+            "h2: P0(100.0,100.0) P25(102.5,102.5) P50(105.0,105.0) P75(107.5,107.5) "
+            "P90(109.0,109.0) P95(109.5,109.5) P99(109.9,109.9) P99.5(109.95,109.95) "
+            "P99.9(109.99,109.99) P100(110.0,110.0)\n",
+            data.toString());
+
+  shutdownThreading();
+}
+
+TEST_P(AdminStatsTest, HandlerStatsJson) {
+  std::string url = "/stats?format=json";
+  Http::TestResponseHeaderMapImpl response_headers;
+  Buffer::OwnedImpl data;
+  MockAdminStream admin_stream;
+  Configuration::MockStatsConfig stats_config;
+  EXPECT_CALL(stats_config, flushOnAdmin()).WillRepeatedly(testing::Return(false));
+  MockInstance instance;
+  store_->initializeThreading(main_thread_dispatcher_, tls_);
+  EXPECT_CALL(instance, stats()).WillRepeatedly(testing::ReturnRef(*store_));
+  EXPECT_CALL(instance, statsConfig()).WillRepeatedly(testing::ReturnRef(stats_config));
+  StatsHandler handler(instance);
+
+  Stats::Counter& c1 = store_->counterFromString("c1");
+  Stats::Counter& c2 = store_->counterFromString("c2");
+
+  c1.add(10);
+  c2.add(20);
+
+  Stats::TextReadout& t = store_->textReadoutFromString("t");
+  t.set("hello world");
+
+  Stats::Histogram& h1 = store_->histogramFromString("h1", Stats::Histogram::Unit::Unspecified);
+  Stats::Histogram& h2 = store_->histogramFromString("h2", Stats::Histogram::Unit::Unspecified);
+
+  EXPECT_CALL(sink_, onHistogramComplete(Ref(h1), 200));
+  h1.recordValue(200);
+
+  EXPECT_CALL(sink_, onHistogramComplete(Ref(h2), 100));
+  h2.recordValue(100);
+
+  store_->mergeHistograms([]() -> void {});
+
+  Http::Code code = handler.handlerStats(url, response_headers, data, admin_stream);
+  EXPECT_EQ(Http::Code::OK, code);
+
+  const std::string expected_json =
+      R"EOF({"stats":[{"value":"hello world","name":"t"},{"value":10,"name":"c1"},{"name":"c2","value":20},{"histograms":{"computed_quantiles":[{"name":"h1","values":[{"cumulative":200,"interval":200},{"cumulative":202.5,"interval":202.5},{"interval":205,"cumulative":205},{"interval":207.5,"cumulative":207.5},{"cumulative":209,"interval":209},{"interval":209.5,"cumulative":209.5},{"interval":209.9,"cumulative":209.9},{"cumulative":209.95,"interval":209.95},{"cumulative":209.99,"interval":209.99},{"cumulative":210,"interval":210}]},{"name":"h2","values":[{"interval":100,"cumulative":100},{"cumulative":102.5,"interval":102.5},{"cumulative":105,"interval":105},{"cumulative":107.5,"interval":107.5},{"interval":109,"cumulative":109},{"cumulative":109.5,"interval":109.5},{"interval":109.9,"cumulative":109.9},{"cumulative":109.95,"interval":109.95},{"interval":109.99,"cumulative":109.99},{"cumulative":110,"interval":110}]}],"supported_quantiles":[0,25,50,75,90,95,99,99.5,99.9,100]}}]})EOF";
```

Comment:
I'm wondering if there is some instability in this gold result as there was a test failure in CI. For some reason I couldn't dive into the details of the failure in this case.

Not sure whether JsonStringEq would iron out ordering diffs; in general Json order of a vector may be significant.

Also I think this might be easier for a human to read if it was patterned after the text-mode comparison above on line 106, rather than using `R"EOF` in this case.

Reply:
Ah! Ok, I finally figure this out. The issue is that the store's histograms() method returns a vector. But it's order depends on the iteration order of a set.
```
std::vector<ParentHistogramSharedPtr> ThreadLocalStoreImpl::histograms() const {
  std::vector<ParentHistogramSharedPtr> ret;
  Thread::LockGuard lock(hist_mutex_);
  {
    ret.reserve(histogram_set_.size());
    for (const auto& histogram_ptr : histogram_set_) {
      ret.emplace_back(histogram_ptr);
    }
  }

  return ret;
}
```
Since this is all rather irrelevant to the stats handler itself, I worked around this by just having a single histogram.

Good point about the pretty-printed expected out. Done.